### PR TITLE
ISOC-3691 Dependabot alerts backfill 

### PR DESCRIPTION
**Pull request by Harminder84**
**What's in this PR?**
Adding support to backfill the dependabot alerts along with the other entities. The backfill behavior is in sync with the backfill for existing entities.

https://hello.atlassian.net/wiki/spaces/CDX/pages/2748121866/LDR+How+do+we+backfill+vulnerabilities+for+existing+GitHub+for+Jira+customers

**Why**
Part of the work to integrate GitHub with Security in Jira
https://hello.atlassian.net/wiki/spaces/CDX/pages/2673460294/RFC+GitHub+for+Jira+Security

**Added feature flags**
ENABLE_GITHUB_SECURITY_IN_JIRA

**Affected issues**  
[ISOC-3691]

**How has this been tested?**  
Unit test

[ISOC-3691]: https://softwareteams.atlassian.net/browse/ISOC-3691

### CODE REVIEW

File: src/sync/dependabot-alerts.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+			schemaVersion: "1.0",
+			id: `d-${transformRepositoryId(alert.repository.id, githubClientConfig.baseUrl)}-${alert.number}`,
+			updateSequenceNumber: Date.now(),
+			containerId: transformRepositoryId(alert.repository.id,  githubClientConfig.baseUrl),
+			displayName: alert.securityAdvisory.summary,
+			description: alert.securityAdvisory.description,
+			url: `${alert.repository.url}/security/dependabot/${alert.number}`,
+			type: "sca",
+			introducedDate: alert.createdAt,
+			lastUpdated: alert.fixedAt || alert.dismissedAt || alert.createdAt,
```
Comment (mjordan-atlassian) [inline src/sync/dependabot-alerts.ts:80]: I think we should also include `alert.autoDismissedAt` here as well incase users have dependabot auto dismissal configured.

File: src/github/client/github-installation-client.ts [hunk 2]
```diff
@@ -362,6 +364,20 @@ export class GitHubInstallationClient extends GitHubClient {
...
+
+		const response = await this.graphql<getDependabotAlertsResponse>(getDependabotAlerts,
+			await this.installationAuthenticationHeaders(),
+			{
+				owner,
+				repo: repoName,
+				per_page: perPage,
+				cursor
+			},
+			{ graphQuery: "getDeploymentsQuery" });
```
Comment (mjordan-atlassian) [inline src/github/client/github-installation-client.ts:377]: Shouldn't this be `getDependabotAlerts`?

File: src/sync/dependabot-alerts.ts (hunk 1 above)
Comment (Harminder84) [reply src/sync/dependabot-alerts.ts]: Included `alert.autoDismissedAt`

File: src/github/client/github-installation-client.ts (hunk 2 above)
Comment (Harminder84) [reply src/github/client/github-installation-client.ts]: Its a metric name. It should be `getDependabotAlerts`. Updated it.

File: src/github/client/github-queries.ts [hunk 3]
```diff
@@ -415,6 +415,82 @@ export type getBranchesResponse = {
 	}
 };
 
+export type getDependabotAlertsResponse = {
```
Comment (rachellerathbone) [inline src/github/client/github-queries.ts:418]: nit: Typescript convention is to use Pascal case for types

File: src/sync/dependabot-alerts.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	const edgeCountEarlierThanFromDate = edges.filter(edge => {
+		const edgeCreatedAt = new Date(edge.node.createdAt);
+		return edgeCreatedAt.getTime() < fromDate.getTime();
+	}).length;
+	return edgeCountEarlierThanFromDate === edges.length;
+};
+
+const transformDependabotAlerts = async (alerts: VulnerabilityAlertNode["node"][], jiraHost: string, logger: Logger, gitHubAppId: number | undefined): Promise<JiraVulnerabilityBulkSubmitData> => {
+
+	const githubClientConfig = await getGitHubClientConfigFromAppId(gitHubAppId, jiraHost);
```
Comment (rachellerathbone) [inline src/sync/dependabot-alerts.ts:67]: githubClientConfig > gitHubClientConfig

File: src/sync/dependabot-alerts.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+const areAllEdgesEarlierThanFromDate = (edges: VulnerabilityAlertNode[], fromDate: Date | undefined) => {
+	if (!fromDate) return false;
+	const edgeCountEarlierThanFromDate = edges.filter(edge => {
+		const edgeCreatedAt = new Date(edge.node.createdAt);
+		return edgeCreatedAt.getTime() < fromDate.getTime();
+	}).length;
+	return edgeCountEarlierThanFromDate === edges.length;
+};
+
+const transformDependabotAlerts = async (alerts: VulnerabilityAlertNode["node"][], jiraHost: string, logger: Logger, gitHubAppId: number | undefined): Promise<JiraVulnerabilityBulkSubmitData> => {
```
Comment (rachellerathbone) [inline src/sync/dependabot-alerts.ts:65]: Can we add a line break here? This line is pretty long.

File: src/sync/scheduler.test.ts [hunk 6]
```diff
@@ -241,4 +241,34 @@ describe("scheduler", () => {
 			expect(task.task).toEqual("commit");
 		});
 	});
+	it("should not filter by dependabot alerts task if ENABLE_GITHUB_SECURITY_IN_JIRA FF is off", async () => {
+		when(booleanFlag).calledWith(BooleanFlags.ENABLE_GITHUB_SECURITY_IN_JIRA, expect.anything()).mockResolvedValue(false);
+		configureRateLimit(10000, 10000);
+		const repoSyncStats = await RepoSyncState.findAllFromSubscription(subscription);
```
Comment (rachellerathbone) [inline src/sync/scheduler.test.ts:247]: Is this a typo? Is it meant to be repoSyncStates?

File: src/github/client/github-installation-client.ts [hunk 7]
```diff
@@ -362,6 +364,20 @@ export class GitHubInstallationClient extends GitHubClient {
 		return response?.data?.data?.viewer?.repositories?.totalCount;
 	}
 
+	public async getDependabotAlertsPage(owner: string, repoName: string, perPage = 1, cursor?: string): Promise<getDependabotAlertsResponse> {
```
Comment (gxueatlassian) [inline src/github/client/github-installation-client.ts:367]: Why `perPage` is 1? Wouldn't that be super slow?

File: src/models/reposyncstate.ts [hunk 8]
```diff
@@ -34,17 +34,20 @@ export interface RepoSyncStateProperties {
 	pullStatus?: TaskStatus;
 	buildStatus?: TaskStatus;
 	deploymentStatus?: TaskStatus;
+	dependabotAlertStatus?: TaskStatus;
```
Comment (gxueatlassian) [inline src/models/reposyncstate.ts:37]: Have you run the db migration scripts yet?

File: src/sync/scheduler.ts [hunk 9]
```diff
@@ -169,7 +170,10 @@ export const getNextTasks = async (subscription: Subscription, targetTasks: Task
 		};
 	}
 
-	const tasks = getTargetTasks(targetTasks);
+	let tasks = getTargetTasks(targetTasks);
```
Comment (gxueatlassian) [inline src/sync/scheduler.ts:173]: Please also update the `jira-sync-post.ts` for bellow

```
const determineSyncTypeAndTargetTasks = async (syncTypeFromReq: string, subscription: Subscription): Promise<SyncTypeAndTargetTasks> => {

	if (syncTypeFromReq === "full") {
		return { syncType: "full", targetTasks: undefined };
	}

	if (subscription.syncStatus === "FAILED") {
		return { syncType: "full", targetTasks: undefined };
	}

	return { syncType: "partial", targetTasks: ["pull", "branch", "commit", "build", "deployment"] };
};

```

File: src/github/client/github-installation-client.ts (hunk 7 above)
Comment (Harminder84) [reply src/github/client/github-installation-client.ts]: It is the default value, same value used in getBranches as well. Updated to 20 now.

File: src/models/reposyncstate.ts (hunk 8 above)
Comment (Harminder84) [reply src/models/reposyncstate.ts:37]: Yes

File: src/github/client/github-installation-client.ts (hunk 7 above)
Comment (Harminder84) [reply src/github/client/github-installation-client.ts]: This is how actual page size calculated
```
const nPages = Math.min(
				// "|| 20" is purely to simplify testing and avoid mocking it all the time
				await numberFlag(NumberFlags.BACKFILL_PAGE_SIZE, 20, jiraHost) || 20,
				// The majority of GitHub APIs hard-limit the page size to 100
				100
			);
```

File: src/sync/scheduler.ts (hunk 9 above)
Comment (Harminder84) [reply src/sync/scheduler.ts:173]: Updated

**Pull request merged**